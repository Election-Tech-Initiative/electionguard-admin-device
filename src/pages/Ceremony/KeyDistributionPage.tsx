import React, { useContext } from 'react'
import { RouteComponentProps, withRouter } from 'react-router-dom'
import styled from 'styled-components'
import CeremonyContext from '../../contexts/ceremonyContext'
import Main, { MainChild } from '../../components/Main'
import Prose from '../../components/Prose'
import Screen from '../../components/Screen'
import Sidebar from '../../components/Sidebar'
import StatusButton, { StatusButtonGrid } from '../../components/StatusButton'
import LinkButton from '../../components/LinkButton'
import { CompletionStatus } from '../../config/types'
import { KEY_ICON, CHECK_ICON } from '../../config/globals'
import SidebarFooter from '../../components/SidebarFooter'

const Header = styled.div`
  margin: 0 auto;
  text-align: center;
`

const KeyDistributionPage = (props: RouteComponentProps) => {
  const { keyVault } = useContext(CeremonyContext)

  const allKeys = Object.keys(keyVault).map(key => keyVault[key])
  const allClaimed = allKeys.reduce(
    (claimed, key) => claimed && key.status === CompletionStatus.Complete,
    true
  )

  const handlePress = (id: string, status: CompletionStatus) => {
    const { history } = props
    if (status !== CompletionStatus.Complete) {
      history.push(`/keys/${id}`)
    }
  }

  const getIcon = (status: CompletionStatus) => {
    return status === CompletionStatus.Complete ? CHECK_ICON : KEY_ICON
  }

  const getStatus = (status: CompletionStatus) => {
    return status === CompletionStatus.Complete ? 'Claimed' : 'Unclaimed'
  }

  return (
    <Screen>
      <Main>
        <MainChild>
          <Prose id="audiofocus">
            <Header>
              <h1>Distribute Keys</h1>
            </Header>
          </Prose>
          <Prose textCenter>
            <p aria-label="Select unclaimed key to distribute">
              <b>Select key to distribute </b>
            </p>
            <StatusButtonGrid>
              {allKeys.map(key => {
                return (
                  <StatusButton
                    key={key.id}
                    onPress={() => handlePress(key.id, key.status)}
                    icon={getIcon(key.status)}
                    statusLabel={getStatus(key.status)}
                    actionLabel="Claim"
                    status={key.status}
                  />
                )
              })}
            </StatusButtonGrid>
          </Prose>
        </MainChild>
      </Main>
      <Sidebar footer={<SidebarFooter />}>
        <p>
          <LinkButton
            big
            primary={allClaimed}
            disabled={!allClaimed}
            to="/setup-encrypters"
            id="next"
          >
            Next
          </LinkButton>
        </p>
        <p>
          <LinkButton disabled small to="/setup-keys" id="back">
            Back
          </LinkButton>
        </p>
      </Sidebar>
    </Screen>
  )
}

export default withRouter(KeyDistributionPage)
