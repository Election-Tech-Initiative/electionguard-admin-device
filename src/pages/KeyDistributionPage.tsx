import React, { useContext } from 'react'
import styled from 'styled-components'
import ElectionContext from '../contexts/electionContext'
import Main, { MainChild } from '../components/Main'
import Prose from '../components/Prose'
import Screen from '../components/Screen'
import ElectionInfo from '../components/ElectionInfo'
import Sidebar from '../components/Sidebar'
import ClaimButton from '../components/ClaimButton'
import LinkButton from '../components/LinkButton'
import { ClaimStatus } from '../config/types'

const Header = styled.div`
  margin: 0 auto;
  text-align: center;
`

const LogoImage = styled.img`
  display: flex;
  margin: 0 auto;
  max-width: 12rem;
`

const KeysGrid = styled.div`
  display: grid;
  grid-auto-rows: minmax(auto, 1fr);
  grid-gap: 0.5rem;
`

const KeyDistributionPage = () => {
  const { election, keyVault } = useContext(ElectionContext)

  const allKeys = Object.keys(keyVault).map(key => keyVault[key])
  const allClaimed = allKeys.reduce(
    (claimed, key) => claimed && key.status === ClaimStatus.Claimed,
    true
  )

  return (
    <Screen>
      <Main>
        <MainChild>
          <Prose id="audiofocus">
            <Header>
              <h1>Distribute Keys</h1>
            </Header>
          </Prose>
          <Prose textCenter>
            <p aria-label="Select unclaimed key to distribute">
              <b>Select key to distribute </b>
            </p>
            <KeysGrid>
              {allKeys.map(key => {
                return (
                  <ClaimButton
                    key={key.id}
                    claimId={key.id}
                    status={key.status}
                    to="/keys"
                  />
                )
              })}
            </KeysGrid>
          </Prose>
        </MainChild>
      </Main>
      <Sidebar
        footer={
          <>
            <hr />
            <ElectionInfo election={election} precinctId="" horizontal />
            <hr />
            <LogoImage
              alt="Election Guard Logo"
              src="/images/electionguard.svg"
            />
          </>
        }
      >
        <p>
          <LinkButton
            big
            primary={allClaimed}
            disabled={!allClaimed}
            to="/setup-encrypters"
            id="next"
          >
            Next
          </LinkButton>
        </p>
        <p>
          <LinkButton disabled small to="/setup-keys" id="back">
            Back
          </LinkButton>
        </p>
      </Sidebar>
    </Screen>
  )
}

export default KeyDistributionPage
