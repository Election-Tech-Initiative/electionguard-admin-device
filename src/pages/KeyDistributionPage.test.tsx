import React from 'react'
import { Route } from 'react-router-dom'

import { render } from '../../test/testUtils'

import KeyDistributionPage from './KeyDistributionPage'

it('renders KeyDistributionPage', async () => {
  const { container } = render(
    <Route path="/" component={KeyDistributionPage} />,
    {
      route: '/',
    }
  )
  expect(container.firstChild).toMatchSnapshot()
})
