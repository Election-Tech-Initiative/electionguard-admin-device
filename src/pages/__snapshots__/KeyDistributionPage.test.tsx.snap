// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders KeyDistributionPage 1`] = `
<div
  class="sc-bxivhb duQixI"
>
  <main
    class="sc-bdVaJa iiFDUH"
  >
    <div
      class="sc-bwzfXH cJzayE"
    >
      <div
        class="sc-htpNat feTyaj"
        id="audiofocus"
      >
        <div
          class="sc-kAzzGY leSgAS"
        >
          <h1>
            Distribute Keys
          </h1>
        </div>
      </div>
      <div
        class="sc-htpNat chxRAJ"
      >
        <p
          aria-label="Select unclaimed key to distribute"
        >
          <b>
            Select key to distribute 
          </b>
        </p>
        <div
          class="sc-kgoBCf hhFZrY"
        />
      </div>
    </div>
  </main>
  <nav
    class="sc-gZMcBi gyqVoq"
  >
    <div
      class="sc-VigVT JNuqd"
    >
      <p>
        <button
          class="sc-jzJRlG byCwxn"
          id="next"
          role="option"
          type="button"
        >
          Next
        </button>
      </p>
      <p>
        <button
          class="sc-jzJRlG fVHyVV"
          disabled=""
          id="back"
          role="option"
          type="button"
        >
          Back
        </button>
      </p>
    </div>
    <div
      class="sc-jTzLTM iBwhQj"
    >
      <hr />
      <div
        aria-hidden="true"
        class="sc-dnqmqq exjFso"
        data-testid="election-info"
      >
        <div
          class="sc-iwsKbI eyMKeV"
        >
          <div
            aria-hidden="true"
            class="sc-ifAKCX jpWdgG"
          />
          <div
            class="sc-htpNat jFlhLj"
          >
            <h5
              aria-label=" General Election."
            >
               General Election
            </h5>
            <p
              class="sc-bZQynM TVUTw"
            >
              November 3, 2020
              <br />
              <span
                class="sc-gzVnrw djKZhV"
              >
                Franklin County
                ,
              </span>
               
              <span
                class="sc-gzVnrw djKZhV"
              >
                State of Hamilton
              </span>
            </p>
            <p
              class="sc-bZQynM hVWfdn"
            >
               
            </p>
          </div>
        </div>
      </div>
      <hr />
      <img
        alt="Election Guard Logo"
        class="sc-chPdSV cFhLDJ"
        src="/images/electionguard.svg"
      />
    </div>
  </nav>
</div>
`;
